apollo-ios-pagination/Sources/ApolloPagination/GraphQLQueryPager.swift

```diff
@@ -1,342 +1,207 @@
-    public func refetch(cachePolicy: CachePolicy = .fetchIgnoringCacheData) {
-      assert(firstPageWatcher != nil, "To create consistent product behaviors, calling `fetch` before calling `refetch` will use cached data while still refreshing.")
-      cancel()
-      fetch(cachePolicy: cachePolicy)
-    }
-
-    public func fetch() {
-      cancel()
-      fetch(cachePolicy: .returnCacheDataAndFetch)
-    }
-
-    /// Cancel any in progress fetching operations and unsubscribe from the store.
-    public func cancel() {
-      nextPageWatchers.forEach { $0.cancel() }
-      nextPageWatchers = []
-      firstPageWatcher?.cancel()
-      firstPageWatcher = nil
-
-      varMap = [:]
-      initialPageResult = nil
-      activeTask?.cancel()
-      activeTask = nil
-      subscribers.forEach { $0.cancel() }
-      subscribers.removeAll()
-    }
+private actor Subscriptions {
+  var subscriptions: Set<AnyCancellable> = []
 
-    /// Whether or not we can load more information based on the current page.
-    public var canLoadNext: Bool {
-      currentPageInfo?.canLoadMore ?? false
-    }
+  func store(subscription: AnyCancellable) {
+    subscriptions.insert(subscription)
+  }
+}
 
-    // MARK: - Private
+private class Completion {
+  var completion: ((Error?) -> Void)?
 
-    private func fetch(cachePolicy: CachePolicy = .returnCacheDataAndFetch) {
-      if firstPageWatcher == nil {
-        firstPageWatcher = GraphQLQueryWatcher(
-          client: client,
-          query: initialQuery,
-          resultHandler: { [weak self] result in
-            guard let self else { return }
-            Task {
-              await self.onInitialFetch(result: result)
-            }
-          }
-        )
-      }
-      firstPageWatcher?.refetch(cachePolicy: cachePolicy)
-    }
+  init(completion: ((Error?) -> Void)?) {
+    self.completion = completion
+  }
 
-    private func onInitialFetch(result: Result<GraphQLResult<InitialQuery.Data>, Error>) {
-      switch result {
-      case .success(let data):
-        initialPageResult = data.data
-        guard let firstPageData = data.data else { return }
-        if let latest {
-          let (_, nextPage) = latest
-          currentValue = .success((firstPageData, nextPage, data.source == .cache ? .cache : .fetch))
-        }
-      case .failure(let error):
-        currentValue = .failure(error)
-      }
+  func execute(error: Error?) async {
+    await MainActor.run { [weak self] in
+      self?.completion?(error)
+      self?.completion = nil
     }
+  }
+}
 
-    private func onSubsequentFetch(
-      cachePolicy: CachePolicy,
-      result: Result<GraphQLResult<PaginatedQuery.Data>, Error>,
-      publisher: CurrentValueSubject<Void, Never>,
-      query: PaginatedQuery
-    ) {
-      switch result {
-      case .success(let data):
-        guard let nextPageData = data.data else {
-          publisher.send(completion: .finished)
-          return
-        }
+private actor CompletionManager {
```

Comment:
Same question I had about the `Subscriptions` actor. What is the value in having 3 different actors here, and does it open us the possiblity of race conditions?

We should probably have some inline documentation about the reasoning here on this for posterity.